src/dune_engine/cached_digest.ml

```diff
@@ -94,19 +94,26 @@ let delete_very_recent_entries () =
   let now = get_current_filesystem_time () in
   match Float.compare cache.max_timestamp now with
   | Lt -> ()
-  | Eq | Gt ->
-    Path.Table.filteri_inplace cache.table ~f:(fun ~key:path ~data ->
-        match Float.compare data.stats.mtime now with
-        | Lt -> true
-        | Gt | Eq ->
-          if !Clflags.debug_digests then
-            Console.print
-              [ Pp.textf
-                  "Dropping cached digest for %s because it has exactly the \
-                   same mtime as the file system clock."
-                  (Path.to_string_maybe_quoted path)
-              ];
-          false)
+  | Eq | Gt -> (
+    let filter (data : file) =
+      match Float.compare data.stats.mtime now with
+      | Lt -> true
+      | Gt | Eq -> false
+    in
+    match !Clflags.debug_digests with
+    | false -> Path.Table.filter_inplace cache.table ~f:filter
+    | true ->
+      Path.Table.filteri_inplace cache.table ~f:(fun ~key:path ~data ->
+          let filter = filter data in
+          if filter then
```

Comment:
Did you mean `not filter`?

I'm not quite sure why we're actually doing this filtering. Do you remember?

Reply:
> Did you mean not filter?

Yes indeed. Thanks.

> I'm not quite sure why we're actually doing this filtering. Do you remember?

Serializing and deserializing aren't cheap and hopefully this will shrink the table before do dump the table on disk.

Reply:
> Serializing and deserializing aren't cheap and hopefully this will shrink the table before do dump the table on disk.

I don't get it. Why are fresh digests less valuable than old ones?

Reply:
You're right, I have no idea. I looked through git blame and it seems like Arseniy introduced this. Might be worth asking him.

Reply:
Ack, I'll ask him tomorrow.

Reply:
So, here is the reason: we can't trust cached digest where the timestamp of the digest coincides with the current time (according to file-system's resolution granularity), so, it's important to filter out such measurements (and it was a concern on some file systems with bad granularity in the past).

Reply:
I added a comment about this.